### ISSUE TITLE
Improve performance of `SlidingSyncRoom::latest_event`

### ISSUE THREAD

**Issue by bnjbvr** [matrix-org/matrix-rust-sdk#1914]
This will create a timeline builder that will do per-event computation, then drop most of the work done to only keep the last event. I like the idea of reusing the timeline builder API, but we should only feed it with one event, the last one.

**Comment by Hywan** [comment]
Some pre-context and other reasons why we should optimise this, https://matrix.to/#/!TAoweeYCttBnjuMqRg%3Amatrix.org/%2440bmAxhSjmldEh6rEQjOYIC89cIyU9ILp-CHDR103yQ

**Comment by pixlwave** [comment]
> we should only feed it with one event, the last one

Not wanting to wade in without too much context, but thought it was worth doubling check here. If the last event isn't a renderable timeline item (e.g. a reaction), will we feed it with more until a timeline item is obtained?

**Comment by jplatte** [comment]
Yes, we can't always just take the latest event. It could be reaction, redaction or something else we don't want to display. I think edits to older messages also don't make sense to be shown as the latest message, but at the same time edits to the otherwise latest message should probably act as the latest message. This is why I implemented it the way I did, and once we have the timeline cached, it won't be a perf problem anymore either.

**Comment by bnjbvr** [comment]
This is a good point! To be fair, current `SlidingSyncRoom::latest_event()` does return an optional `EventTimelineItem`, and unless I'm missing something, its `content` can be a message but also something else like a `RedactedMessage` or `Sticker` or `UnableToDecrypt` etc. (basically all variants in `TimelineItemContent` that's not a `Message`).

**Comment by jplatte** [comment]
Right, we should probably return the latest one of the usual timeline item's that not UTD or redacted! I think I wrote `latest_event` before we those types.

**Comment by bnjbvr** [comment]
I think this is fixed by caching the latest event in the `RoomInfo`.
